Use a single threadpool configured via new option :threads

aviflax opened the pull request:
I’ve been debugging slow test suite execution times (wall-clock-time) in
my test suite and I found a few problems:

* When running in CircleCI, `Runtime.availableProcessors` always
  [erroneously returns 32](https://circleci.com/docs/2.0/configuration-reference/#resource_class), when the correct number is generally 2–8.
* When `:multithread` is set to `true` (the default), each namespace is
  tested simultaneously with its own dedicated threadpool, and each
  threadpool uses `(+ 2 (.availableProcessors runtime))` threads.

The combination of these factors means that for a project with, say, 4
namespaces to be tested, when run on CircleCI four threadpools would be
created, each with 34 threads, and they’d all be running tests
simultaneously, which means that 136 threads would be simultaneously
executing tests. If those tests are computationally intensive, as e.g.
property tests tend to be, then that can drastically lower test
throughput on a container that may actually have only 2 or 4 cores
available.

And due to test.check and clojure.spec, such computationally intensive
property tests are becoming more common in the Clojure community.

These changes address these issues like so:

* A single threadpool is always used to run all tests
  * Makes it easier to control and reason about the number of threads
    used to execute tests
  * Even in the CircleCI scenario, only 32 threads total would be used,
    rather than `(* 32 (count namespaces-under-test))`
* A new option `:threads` enables users to specify the number of threads
  they’d like to be used to execute tests
  * So in circumstances such as that in CircleCI, it should be fairly
    easy for users to work around the erroneous value reported by Java
  * The default value is changed to be the value returned by `Runtime`,
    rather than that value plus two. This is because I think that
    computationally intensive test suites are becoming more common in
    the Clojure community, and this default would therefore be a little
    more broadly appropriate, I think. It also just feels more intuitive
    to me, more in line with the principle of least astonishment.

All the tests pass for me, well, except for the test that was already failing in master (see #62) (I also rebased this branch onto that bugfix branch locally and then _all_ the tests passed).

danielcompton commented:
I can see the benefit of a single threadpool, and also setting the max number of threads. I would note that l think Java 9/10 fixes `Runtime#availableProcesors` returning more CPUs than are available to the process. https://blog.docker.com/2018/04/improved-docker-container-integration-with-java-10/ - I'm pretty sure this would work in non-Docker contexts too.

aviflax commented:
@danielcompton 😄 & good to know!

aviflax commented:
@danielcompton FYI, while I don’t doubt that there may be improvements to `Runtime#availableProcessors` in Java 9/10, it appears that in the specific case of a container executing in CircleCI, an erroneous value is still reported:

I restarted a CircleCI build job with SSH access, connected, then ran this:

```clojure
~# clj
Clojure 1.10.0
user=> (System/getProperty "java.version")
"11.0.1"
user=> (.availableProcessors (Runtime/getRuntime))
36
```

To be clear, I don’t blame _Java_ for this — as far as I can tell, it seems to be a flaw in CircleCI. Java is apparently reading the value from somewhere under `/proc` and that “file” is returning an erroneous value.

FWIW, before I worked on this proposed change to eftest I did some experimentation with the new(ish) ([added in Java 10](https://bugs.openjdk.java.net/browse/JDK-8146115)) JVM flag `-XX:ActiveProcessorCount=nn` to confirm that when `Runtime#availableProcessors` reports a more reasonable number, eftest’s behavior was more reasonable. The results of those experiments are what encouraged me to propose these changes to eftest itself. While I _could_ influence the number of threads used by eftest by specifying that JVM flag when using `clojure` to run my test suite, that would be a highly baroque, obscure, indirect, and headache-inducing approach. Hence these proposed changes.

vemv commented:
Nice work!

IMO (note: not the author here), the best level of granularity is not `:threads` but `:executor`. i.e., let consumers specify whatever executor they please.

Otherwise you are changing one fixed strategy for another fixed strategy, which predictably won't leave everyone happy.

Also worth noting that some test suites are I/O bound (e.g. Selenium ones), so a strategy that revolves around CPU core count might be excessive (or inefficient).

aviflax commented:
> Nice work!

Thanks! 

> IMO (note: not the author here), the best level of granularity is not `:threads` but `:executor`. i.e., let consumers specify whatever executor they please.
> 
> Otherwise you are changing one fixed strategy for another fixed strategy, which predictably won't leave everyone happy.
> 
> Also worth noting that some test suites are I/O bound (e.g. Selenium ones), so a strategy that revolves around CPU core count might be excessive (or inefficient).

Very good points, thank you!

I’m thinking that another way of looking at this change is that I’m changing the current fixed strategy by making it a little less fixed — by making the number of threads configurable. This fits my personal approach to software evolution, which is to try to make it as incremental as possible. It seems to me that this change, in its current form, would be a good step towards the scenario you’ve described; now that there’s only a single Executor created and used for each invocation of `test-all` it’ll be easier to reason about the resource utilization of a given Executor for a given test suite.

Also, I’m guessing you have in mind that the value of the `:executor` key would be an Executor object, and while that might be a reasonable scenario to support, I like that right now the options hash is just data, and I’d like to continue to support that approach.  If we’re going to support various kinds of executors, I’d prefer that a user just be able to specify values like `:threadpool` or `:forkjoin` — I don’t want someone new (or new-ish) to Clojure to _have_ to learn Java interop in order to specify which Executor they want to use. And if we do support such pre-baked options, well then the `:threads` option will continue to be useful and relevant, when the `:threadpool` option is used (which I assume would be the default, for backwards compatibility).

Does that make any sense?

vemv commented:
> I’m thinking that another way of looking at this change is that I’m changing the current fixed strategy by making it a little less fixed — by making the number of threads configurable.

I concur!

>  I like that right now the options hash is just data

I see the value of this, particularly when invoking this from Leiningen, a scenario where the last thing one would expect is having to write some Executor.

> I’d prefer that a user just be able to specify values like :threadpool or :forkjoin

This is a sweet spot between your and my proposal: eftest internals are concerned with executor objects, and (most) users do only interact data.

As a last observation, executors could potentially give some unforeseen capabilities to eftest. For example, distributing the workload across multiple computers, e.g. having a local Android farm for device testing. In the end the Executor interface is extremely simple (and not concerned with threads at all), allowing one run code as desired.

aviflax commented:
> I see the value of this, particularly when invoking this from Leiningen, a scenario where the last thing one would expect is having to write some Executor.

Yes, exactly!

> This is a sweet spot between your and my proposal: eftest internals are concerned with executor objects, and (most) users do only interact data.

+1

> As a last observation, executors could potentially give some unforeseen capabilities to eftest. For example, distributing the workload across multiple computers, e.g. having a local Android farm for device testing. In the end the Executor interface is extremely simple (and not concerned with threads at all), allowing one run code as desired.

This is compelling!

Just to be super clear though, I’d still like to leave this PR as-is for now, in terms of approach and scope (unless and until we hear otherwise from @weavejester of course). The additional improvements we’ve been discussing would make a lot of sense as follow-up PRs.

weavejester commented on README.md:
```diff
@@ -94,6 +94,27 @@ Or you can synchronize the entire namespace:
+##### Setting the number of threads used
+
+When multithreading is enabled, Eftest uses a single fixed-threadpool
+[`ExecutorService`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/ExecutorService.html)
```
Links with long URLs like this should be defined as reference-style links.

weavejester commented on README.md:
```diff
@@ -94,6 +94,27 @@ Or you can synchronize the entire namespace:
+
+By default, Eftest will instantiate the threadpool with the number of
+processors (cores) available to the JVM, as reported by
+<a href="https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Runtime.html#availableProcessors()"><code>Runtime.availableProcessors</code></a>.
```
This HTML code should be replaced with markdown.

weavejester commented on README.md:
```diff
@@ -94,6 +94,27 @@ Or you can synchronize the entire namespace:
+processors (cores) available to the JVM, as reported by
+<a href="https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Runtime.html#availableProcessors()"><code>Runtime.availableProcessors</code></a>.
+(NB: in some circumstances, such as
+[in a CircleCI test container](https://circleci.com/docs/2.0/configuration-reference/#resource_class),
```
This link should be reference-style.

weavejester commented on README.md:
````diff
@@ -169,19 +190,21 @@ logging long tests.
 ```
 
-### Plugin
+### Leiningen Plugin
````
Only include changes related to the feature being added.

weavejester commented on README.md:
```diff
@@ -169,19 +190,21 @@ logging long tests.
 
-You can customize the reporter and set the concurrency by adding an
-`:eftest` key to your project map:
+You can customize the reporter, set the concurrency mode, set the
```
Change this to:

    You can customize the reporter and configure the concurrency settings
    by adding an `:eftest` key to your project map:

weavejester commented on eftest/src/eftest/runner.clj:
```diff
@@ -44,22 +44,22 @@
-(defn- pmap* [f xs]
-  (pcalls* (map (fn [x] #(f x)) xs)))
+(defn- ^ExecutorService threadpool-executor
+  [{:keys [threads] :or {threads :processors}}]
```
Lets remove the `:processors` option, and instead set the default to 2 + available-processors when the number of threads is not set.

weavejester commented on eftest/src/eftest/runner.clj:
```diff
@@ -44,22 +44,22 @@
+      (.availableProcessors (Runtime/getRuntime)))))
+
+(defn- pcalls* [executor fs]
+   (->> fs
```
Indentation needs to be fixed.

weavejester commented on eftest/src/eftest/runner.clj:
```diff
@@ -169,6 +175,12 @@
                        in those namespaces are run serially. If set to :vars,
                        the namespaces are run serially, but the vars inside run
                        in parallel.
+    :threads         - either a positive integer or :processors (defaults to
```
Remove the bit about processors. This can just be a brief description.

weavejester commented on eftest/src/eftest/runner.clj:
```diff
@@ -99,29 +99,35 @@
           (test/do-report (fixture-exception t)))))))
 
-(defn- test-ns [ns vars report opts]
+(defn- test-ns [ns vars executor report opts]
```
The executor can be passed through `opts`.

weavejester commented on eftest/src/eftest/runner.clj:
```diff
@@ -99,29 +99,35 @@
 (defn- test-all [vars {:as opts :keys [capture-output?] :or {capture-output? true}}]
   (let [report (synchronize test/report)
-        mapf   (if (multithread-namespaces? opts) pmap* map)
+        executor (threadpool-executor opts)
```
Align let forms, and it's probably better to define the executor in `run-tests`, considering that `:executor` might be an option in future.

aviflax commented:
> Thanks for the PR. I've reviewed and added some comments.

My pleasure, thanks for the detailed review! I’ll make these changes soon.

aviflax commented:
@weavejester I believe I’ve addressed all your feedback in 05a419d. The tests all still pass (when master is merged in, so as to get the fix in #62). Thanks for the thorough and clear review!

weavejester commented on eftest/src/eftest/runner.clj:
```diff
@@ -44,22 +44,23 @@
-(defn- pmap* [f xs]
-  (pcalls* (map (fn [x] #(f x)) xs)))
+(defn- default-threads
+  []
```
No need to put the argument list on its own line.

weavejester commented on eftest/src/eftest/runner.clj:
```diff
@@ -182,7 +191,10 @@
          (test/do-report {:type :begin-test-run, :count (count vars)})
-         (let [counters (test-all vars opts)
+         (let [executor (threadpool-executor opts)
+               opts′    (assoc opts :executor executor)
```
That tick at the end of `opts′` doesn't look like it's ASCII. I don't think you'd need a new variable name here either; just use `opts`.

aviflax replied:
It isn’t! I didn’t know you preferred ASCII-only; fixing this now. (I personally like that Clojure supports “real” characters and I like to use them in my own code, but I understand that’s not for everyone!) As for naming it just `opts` — sure, I’ll do that. Personally I prefer to avoid shadowing, but I’ll go with your preference here. Thanks!

weavejester commented on eftest/src/eftest/runner.clj:
```diff
@@ -182,7 +191,10 @@
                  test/report      (:report opts progress/report)]
          (test/do-report {:type :begin-test-run, :count (count vars)})
-         (let [counters (test-all vars opts)
+         (let [executor (threadpool-executor opts)
```
We only need  an executor if we're executing threaded tests, otherwise it can be `nil`.

weavejester commented on README.md:
```diff
@@ -94,6 +94,30 @@ Or you can synchronize the entire namespace:
+container][CircleCI-Config-resource_class],
+`Runtime.availableProcessors` returns an erroneous value.)
+
+[ExecutorService-Javadoc]: https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/ExecutorService.html
```
We don't need to be so unambiguous with the labels. As long as they're unique:

```markdown
When multithreading is enabled, Eftest uses a single fixed-threadpool
[`ExecutorService`][executorservice] to run all selected tests.

By default, Eftest will instantiate the threadpool with the number of
processors (cores) available to the JVM, as reported by
[`Runtime.availableProcessors`][availableprocessors].  (Note that in
some environments, such as [in a CircleCI test
container][resource-class], `Runtime.availableProcessors` returns an
erroneous value.)
```

weavejester commented on README.md:
````diff
@@ -94,6 +94,30 @@ Or you can synchronize the entire namespace:
+positive integer:
+
+```clojure
+user=> (run-tests (find-tests "test") {:threads 4})
````
Do you think `:thread-count` would be a less ambiguous name for the option?

aviflax replied:
Yes, but IMHO the current name is already sufficiently unambiguous. I’ll leave this as-is unless you ask me to change it 😅

weavejester replied:
Let's change it to `:thread-count`. I think that works better with the `:multithread?` options.

aviflax commented:
> Some minor comments. Sorry for taking a while to get around to reviewing this.

No worries. I appreciate the review!

I believe I’ve addressed your feedback in 3c6839e — and, I hope you don’t mind, I also added a comment to `default-threads` — let me know if you’d like me to remove it and I’ll do so.

weavejester commented on eftest/src/eftest/runner.clj:
```diff
@@ -44,37 +44,42 @@
-(defn- pmap* [f xs]
-  (pcalls* (map (fn [x] #(f x)) xs)))
+(defn- default-threads []
+  ;; NB: Runtime/getRuntime returns erroneous results in CircleCI; see
```
I think we can remove this comment; it's too specific to a particular platform.

aviflax commented:
> A couple of minor changes, and then I think we're ready to merge.

Done in 7771c20.

aviflax commented:
> Thanks for your changes, and apologies for the delay in getting back to you.

My pleasure, and no worries.

> Can you squash your changes down so there's a clean history?

Sure, done.

(As a rule I avoid force-pushing to branches while they’re being actively reviewed; it makes it harder to review the follow-up changes. Also, just in case you’re not aware, there’s a GitHub repo setting you could enable that allows for a [Squash and Merge](https://help.github.com/articles/about-pull-request-merges/#squash-and-merge-your-pull-request-commits) mode when merging pull requests, so going forward you could easily do the squashing yourself when you’re ready to merge, right here in the PR UI. In addition to reducing the round-trip waiting for a reviewer to perform the squash, it also adds the squashed commit directly to master, without a merge commit — yielding an even cleaner history in master. So even in this case wherein I’ve already squashed the commits in this branch, you might like using this feature.)

> I follow the [seven rules of a great git commit message](https://chris.beams.io/posts/git-commit/#seven-rules) for my projects.

Cool page, thanks for the link!

weavejester commented:
> As a rule I avoid force-pushing to branches while they’re being actively reviewed

Yes, I do too. I think that's perfectly reasonable.

> Also, just in case you’re not aware, there’s a GitHub repo setting you could enable that allows for a [Squash and Merge](https://help.github.com/articles/about-pull-request-merges/#squash-and-merge-your-pull-request-commits) mode when merging pull requests

Unfortunately it's a squash and *fast-forward* merge, which makes it useless for my purposes.

> In addition to reducing the round-trip waiting for a reviewer to perform the squash, it also adds the squashed commit directly to master, without a merge commit — yielding an even cleaner history in master.

That's exactly what prevents me from using it; I want a readable history, but I also don't want to discard relevant information. When a PR is merged I want an explicit merge commit that links back to the PR and gives an overview of the changes made.

weavejester commented:
Can you change the commit message to:

```
Add common threadpool with :thread-count option

When running tests a single threadpool with a configurable thread count
should be used. Previous versions used a threadpool per namespace,
resulting in more threads used than would be performant. Additionally,
on environments that misrepresent the available processors (such as
CircleCI), also result in too many threads being used.
```

The current commit message subject is a shade too long, and the commit body could be trimmed down quite a bit without losing relevance.

aviflax commented:
> Unfortunately it's a squash and _fast-forward_ merge, which makes it useless for my purposes.

Aha, right, yes, good point!

> That's exactly what prevents me from using it; I want a readable history, but I also don't want to discard relevant information. When a PR is merged I want an explicit merge commit that links back to the PR and gives an overview of the changes made.

💯😍💯 

> Can you change the commit message…

Done!

weavejester commented:
Thanks! Merged.

aviflax commented:
Great! My pleasure!

Any chance of a release soon so I can start using this in my main project?

weavejester commented:
Released. Sorry for taking a while to get around to it.

weavejester commented:
Aaaaand I've managed to mess it up. Hang on :)

weavejester commented:
Released as 0.5.6, hopefully correctly this time.

aviflax commented:
Great, thanks!
